src/jit/assertionprop.cpp

```diff
@@ -2243,6 +2246,16 @@ AssertionIndex Compiler::optAssertionIsSubrange(GenTree* tree, var_types toType,
                 continue;
             }
 
+            // If we have an unsigned fromType, then the loBound can't be negative
+            //
+            if (varTypeIsUnsigned(fromType))
+            {
+                if (curAssertion->op2.u2.loBound < 0)
```

Comment:
Speaking of `fromType` signedness - it looks like this condition is unnecessarily strong. For a narrowing cast the signedness of `fromType` doesn't really matter as you actually have a widening cast from `toType`. Yeah, "from `toType`", that makes my head spin :smile: Well, at least this is only a CQ issue in the worst case.

Reply:
When we have a from type of TYP_UINT and a toType of TYP_INT it matters.

By the way is this a narrowing or a widening cast?

I look a the casting operation like this:

resulting type  <= CastToType  <= fromType

Where both the fromType and the resulting type are constrained to be legal IL stack types.
(which are 32-bit or 64-bit  types)

The small types only show up in the CastToType.

Reply:
> When we have a fromType of TYP_UINT and a toType of TYP_INT it matters.

Only if it's an overflow checking cast, otherwise the entire cast is no-op. In general:
* casts from `TYP_(U)INT` to `TYP_(U)LONG` - `GTF_UNSIGNED` decides whether we're sign extending or zero extending
* overflow checking casts - `GTF_UNSIGNED` decides how the source operand is treated for overflow checking purposes (e.g. when casting unsigned to `TYP_SHORT` check for 0..32767 instead of -32768..32767).
* all other integral casts should ignore `GTF_UNSIGNED`. Ideally it should not be set at all but I'm not sure the JIT gets it right, it's too easy to change IR in various ways and forget to reset the flag.

> By the way is this a narrowing or a widening cast?

Hmm, neither I'd say. If it doesn't check for overflow then it's a no-op. If it checks for overflow it's not really a cast in the narrowing/widening sense, it's just a range check.

Casting to a small type can probably be considered both narrowing and widening (`TYP_INT`->`TYP_SHORT`->`TYP_INT`) but I decided a long time ago that my head will spin less if I classify it based on the generated code. It generates a `movsx` so it's a widening cast. Unless of course, it checks for overflow when again it's not a cast but a range check.